### PR TITLE
[Draft] Biswap DEX AMM

### PR DESCRIPTION
**Pull request by fortysevenlabs**
Please review [README](https://github.com/fortysevenlabs/messari-subgraphs/blob/a0d9b236d23e3b4306196b34ea095835e968f1fa/subgraphs/biswap/README.md). There are a few questions in the "Protocol-Side revenue" pieces.

* [Messari Subgraphs Explorer URL](https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/id/QmPGSDt25U6UXgFctLGyVSprzMSoAHpZpHkYDYRXKp3nE5)
* [Graph URL](https://api.thegraph.com/subgraphs/id/QmPGSDt25U6UXgFctLGyVSprzMSoAHpZpHkYDYRXKp3nE5) - Status: Syncing, it's really slow.

Follow ups:
- [ ] Thumbs up on methodology from Messari Team
- [ ] Update deployment config (package.json, deployments)
- [ ] Update development status
- [ ] Deploy subgraph to messari/biswap-bsc
- [ ] Merge PR

Tarun Todo - Issues Identified
- [ ] set `liquidityPoolHourlySnapshots-cumulativeSupplySideRevenueUSD` to not be reset on new entity creation
- [ ] set `liquidityPoolHourlySnapshots-hourlySupplySideRevenueUSD` to not be reset on new entity creation
- [ ] set `liquidityPoolHourlySnapshots-cumulativeTotalRevenueUSD cumulative` to not be reset on new entity creation

### CODE REVIEW

**Comment by fortysevenlabs** [comment]
@this-username-is-taken please assign a reviewer

File: subgraphs/biswap/src/common/creators.ts [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+  poolAddress: string,
+  token0Address: string,
+  token1Address: string
+): void {
+  let protocol = getOrCreateDex();
+
+  // create the tokens and tokentracker
+  let token0 = getOrCreateToken(event, token0Address);
+  let token1 = getOrCreateToken(event, token1Address);
+  let poolToken = getOrCreateLPToken(poolAddress, token0, token1);
```
Comment (fortysevenlabs) [inline subgraphs/biswap/src/common/creators.ts:46]: Are we cool with grabbing the the name, symbol, decimals from the LP token contract directly? (see getOrCreateToken vs getOrCreateLPToken). I think that is better.

Comment (fortysevenlabs) [reply subgraphs/biswap/src/common/creators.ts:46]: Do we want to track the price of the poolToken/outputToken (using prices lib)? I believe the prices lib won't have price for most LP tokens and/or they may be inaccurate.

**Review by steegecs — COMMENTED** [review]
Are you planning to track the Sync and Transfer events? These track useful information like pool reserves and the minting and burning of LP tokens. Also did you consider building this subgraph in with the uniswap-forks? The contracts look quite the same if I am not mistaken.

**Comment by fortysevenlabs** [comment]
I was not considering tracking Sync and Transfer events. AFAIK both sync and transfer events will update the input (LP minted) and output token (LP burned) balances.
* Can add these but curious, wouldn't these be derived anyway through each Mint and Burn event?
* Note that biswap syncing is extremely slow (it has synced 2% over the last day).

I learned about uniswap-forks setup after I created this unfortunately.
* Is it mandatory to have biswap subgraph as part of uniswap-forks?
* I can take a stab at adding biswap there (and assuming it should not be time consuming) but worried about the additional time I may need to spend.

**Comment by fortysevenlabs** [comment]
@steegecs is there a guide on how to run subgraph deploys for uniswap-forks to personal targets from the uniswap-forks directory?

**Comment by fortysevenlabs** [comment]
@steegecs took a stab at adding biswap to uniswap-forks. Take a look and share your thoughts. I'm blocked on how do I run builds and deploys when developing with uniswap-forks?

```
in directory:  messari-subgraphs/subgraphs/uniswap-forks
$ ../../messari-cli/bin/messari build --service hosted-service --target fortysevenlabs
✔ Choose a scope: · protocol
✔ Deployment id to use for the build/deployment · biswap

RESULTS:
Build Failed: biswap-bsc
END
```

**Comment by steegecs** [comment]
> @steegecs is there a guide on how to run subgraph deploys for uniswap-forks to personal targets from the uniswap-forks directory?

There is not a guide, but if you run `messari b` you'll see there is a `--slug` option. If you want to deploy to a target other than the standard messari target specified under the `services` in the deployment.json, you get pass the --slug with a subsequent value to overwrite that location. For the target, you can just specify your account name on the graph hosted service or pass your authorization for subgraph studio.

**Comment by steegecs** [comment]
> @steegecs took a stab at adding biswap to uniswap-forks. Take a look and share your thoughts. I'm blocked on how do I run builds and deploys when developing with uniswap-forks?
> 
> ```
> in directory:  messari-subgraphs/subgraphs/uniswap-forks
> $ ../../messari-cli/bin/messari build --service hosted-service --target fortysevenlabs
> ✔ Choose a scope: · protocol
> ✔ Deployment id to use for the build/deployment · biswap
> 
> RESULTS:
> Build Failed: biswap-bsc
> END
> ```

**Comment by steegecs** [comment]
> I was not considering tracking Sync and Transfer events. AFAIK both sync and transfer events will update the input (LP minted) and output token (LP burned) balances.
> 
> * Can add these but curious, wouldn't these be derived anyway through each Mint and Burn event?
> * Note that biswap syncing is extremely slow (it has synced 2% over the last day).
> 
> I learned about uniswap-forks setup after I created this unfortunately.
> 
> * Is it mandatory to have biswap subgraph as part of uniswap-forks?
> * I can take a stab at adding biswap there (and assuming it should not be time consuming) but worried about the additional time I may need to spend.

**Comment by steegecs** [comment]
> I was not considering tracking Sync and Transfer events. AFAIK both sync and transfer events will update the input (LP minted) and output token (LP burned) balances.
> 
> * Can add these but curious, wouldn't these be derived anyway through each Mint and Burn event?
> * Note that biswap syncing is extremely slow (it has synced 2% over the last day).
> 
> I learned about uniswap-forks setup after I created this unfortunately.
> 
> * Is it mandatory to have biswap subgraph as part of uniswap-forks?
> * I can take a stab at adding biswap there (and assuming it should not be time consuming) but worried about the additional time I may need to spend.

Yeah - I suppose we could try a derivation, and it probably would decrease indexing time by a lot. I or you can give that a try and see if we can get the data to reflect accurately. 

I think it is best to include with the uniswap-forks. Primarily because the code it pretty much the exact same, so it makes sense to reuse. Makes it easy to keep all of these subgraphs in sync as well, and there is already the infrastructure for configuration.

**Comment by steegecs** [comment]
> Yes, all good and makes sense on closing this PR.

I did not mean to close XD.

**Pull request closed without merging**